N°1000 - Portal: filter BrowseBrick result varies if in list or tree

bdalsass opened the pull request:
https://support.combodo.com/pages/UI.php?operation=details&class=Bug&id=1000&

Filtre composant browse brick du portail en mode tree view.
La recherche s'effectue aussi sur les données non visibles dans l'arbre.

![2024-10-30_16h09_19](https://github.com/user-attachments/assets/c96be4ca-c874-491a-94ff-bcd553d521b8)

bdalsass committed:
N°1000 - Portal: filter BrowseBrick result varies if in list or tree

steffunky commented on datamodels/2.x/itop-portal-base/portal/src/Helper/BrowseBrickHelper.php:
```diff
@@ -514,6 +516,7 @@ public function AddToTreeItems(array &$aItems, array $aCurrentRow, array &$aLeve
 				'name' => $aCurrentRowValues[0]->Get($aLevelsProperties[$aCurrentRowKeys[0]]['name_att']),
 				'class' => get_class($aCurrentRowValues[0]),
 				'subitems' => array(),
+				'filter_data' => $this->GetFilterData($aLevelsProperties[$aCurrentRowKeys[0]], $aCurrentRowKeys[0], $aCurrentRowValues[0], '&nbsp;&nbsp●&nbsp&nbsp'),
```
If you commit `&nbsp;` to add spacing between elements 1 more time I hope you run faster than @Molkobain

steffunky replied:
Moreover your `&nbsp` are missing a `;` most of the time

steffunky commented on datamodels/2.x/itop-portal-base/portal/src/Helper/BrowseBrickHelper.php:
```diff
@@ -564,4 +567,56 @@ public function AddToTreeItems(array &$aItems, array $aCurrentRow, array &$aLeve
+
+				// append to result
+				$sValues .= $sValue;
+				$sValuesAndCodes .= '<b>' . $aField['label'] . ':</b> ' . $sValue;
```
It would be better to use a `<span>` combined with a `class` and then style it

steffunky commented on datamodels/2.x/itop-portal-base/portal/public/lib/jquery-treelistfilter/js/TreeListFilter.js:
```diff
@@ -41,6 +41,7 @@ jQuery.fn.treeListFilter = function(list, timeout, callback) {
                     // Modified the search so it looks  for each parts of the search and not the entire sentance
                     // Modified the text to remove accents (latinise())
                     var text = liObject.find('.tree-item-wrapper').text();
+	                text+= liObject.find('.tree-item-wrapper').attr('data-tree-additional-search');
```
We'll need to add a comment in the file header to specify that we altered the library source code

Molkobain replied:
You can also fork it in the _combodo-itop-libs_ organization, that way you can keep track of your changes.

steffunky commented on datamodels/2.x/itop-portal-base/portal/templates/bricks/browse/mode_tree.html.twig:
```diff
@@ -197,10 +200,12 @@
                 // Building node
                 $('ul[data-level-id="'+nodeId+'"]').append(liElem);
                 spanElem.append(nameElem);
+                spanElem.append('<div class="tree-item-filter-data" style="display: none">' + item['filter_data']['values_and_codes'] + '</div>');
```
Use your `tree-item-filter-data` class to handle styles and avoid inline styling

steffunky replied:
@Molkobain @Molkobain @Molkobain @Molkobain @Molkobain @Molkobain  he keeps doing it

Molkobain replied:
![](https://i.giphy.com/F3BeiZNq6VbDwyxzxF.webp)

steffunky commented on datamodels/2.x/itop-portal-base/portal/templates/bricks/browse/mode_tree.html.twig:
```diff
@@ -197,10 +200,12 @@
 
                 var liElem = $('<li></li>').addClass('list-group-item');
                 var spanElem = $('<span></span>').addClass('tree-item-wrapper').attr('data-item-id', item.id).attr('data-level-alias', item.level_alias);
+                spanElem.attr('data-tree-additional-search', item['filter_data']['values']);
```
You can add `.attr(..)` just before `;` on the above line to avoid repeating `spanElem.`

Molkobain commented on datamodels/2.x/itop-portal-base/portal/src/Helper/BrowseBrickHelper.php:
```diff
@@ -514,6 +516,7 @@ public function AddToTreeItems(array &$aItems, array $aCurrentRow, array &$aLeve
 				'name' => $aCurrentRowValues[0]->Get($aLevelsProperties[$aCurrentRowKeys[0]]['name_att']),
 				'class' => get_class($aCurrentRowValues[0]),
 				'subitems' => array(),
+				'filter_data' => $this->GetFilterData($aLevelsProperties[$aCurrentRowKeys[0]], $aCurrentRowKeys[0], $aCurrentRowValues[0], '&nbsp;&nbsp●&nbsp&nbsp'),
```
Please tell me the `&nbsp;` are something you committed by mistake please?
Otherwise, I know where you live.

steffunky replied:
It's bug N°1000, he used the ancient ways in his code

bdalsass replied:
> Please tell me the `&nbsp;` are something you committed by mistake please? Otherwise, I know where you live.

i'am ready

![PXL_20241031_090234717 MP](https://github.com/user-attachments/assets/2e044d37-11fb-4d56-b6a0-e0673d4bd42f)

Molkobain replied:
😂😂😂😂❤️

steffunky commented on datamodels/2.x/itop-portal-base/portal/src/Helper/BrowseBrickHelper.php:
```diff
@@ -564,4 +567,56 @@ public function AddToTreeItems(array &$aItems, array $aCurrentRow, array &$aLeve
+	 * @param array $aLevelProperties tree level properties
+	 * @param string $sRowKey row key
+	 * @param \DBObject $oRowValue row value
+	 * @param string $sValuesAndCodesSeparator separator between values codes pairs
```
Does this default value have something to do with `TreeListFilter` default behavior ? If so I think we should make it clear by using a constant

steffunky replied:
If you only use this variable for spacing and adding a character between them, I think It would be easier to add a class surrounding each of you values and the use sibling selector and add space + your dot or your pipe through CSS

bdalsass committed:
N°1000 - Portal: filter BrowseBrick result varies if in list or tree
- replace html separators by css class

bdalsass committed:
N°1000 - Portal: filter BrowseBrick result varies if in list or tree
- replace html separators by css class

bdalsass committed:
N°1000 - Portal: filter BrowseBrick result varies if in list or tree
- add alteration comment to Tree List Filter jQuery plugin 1.0
- remove display none to tree-item-filter-data

vbkunin commented:
Hi there! @bdalsass @steffunky @Molkobain 
I’m a bit confused — is this a bug or a feature? Could you please clarify?
The documentation says that fields with hidden=true should not be displayed, but still be used for search.
However, the hidden flag seems to be completely ignored now, and all my hidden fields are showing up in the tree.

> `/itop_design/module_designs/module_design/bricks/brick/levels/level/fields/field[@id='key_words']/hidden`
> Hide the field so it won't be displayed but still be used by the filter. Available values true|false, default is false.

vbkunin commented:
Oh, I think I get it now! These fields should only appear after a search is performed — that is, when there’s some text in the search input (probably to help users understand why a particular item shows up in the results). They should hide again when the search field is cleared. https://github.com/Combodo/iTop/pull/673/files#diff-4061e642a80df83b39fd60608ed61e7f45b4cd307f7a868eb6d325f7effabdbaR153
But for some reason, all these fields also appear when the tree is initially loaded and fully expanded using the "+" button. It’s possible that the` .tree-item-filter-data` class was originally meant to have `display: none` style by default.
